File: java/BinaryTreeMaximumPathSum.java
public class BinaryTreeMaximumPathSum {
    private int maxSum;

    public int maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        maxSum = Integer.MIN_VALUE;
        searchMax(root);
        return maxSum;
    }

    private int searchMax(TreeNode curr) {
        if (curr == null) {
            return 0;
        }
        int leftMax = Math.max(0, searchMax(curr.left));
        int rightMax = Math.max(0, searchMax(curr.right));
        maxSum = Math.max(maxSum, curr.val + leftMax + rightMax);
        return curr.val + Math.max(0, Math.max(leftMax, rightMax));
    }
}

File: python/binary_tree_maximum_path_sum.py
#!/usr/bin/env python3

from tree_node import TreeNode

class Solution:
    def __init__(self):
        self.max_sum = -0x3f3f3f3f
    # @param root, a tree node
    # @return an integer
    def maxPathSum(self, root):
        self.helper(root)
        return self.max_sum

    def helper(self, node):
        if not node:
            return 0
        left_sum = self.helper(node.left)
        right_sum = self.helper(node.right)
        t = node.val + max(0, left_sum) + max(0, right_sum)
        if t > self.max_sum:
            self.max_sum = t
        return node.val + max(0, left_sum, right_sum)

def main():
    solver = Solution()

    root = TreeNode(1)
    root.left = TreeNode(2)
    root.left.left = TreeNode(3)
    print(solver.maxPathSum(root))

if __name__ == '__main__':
    main()
